### DISCUSSION TITLE
How to determine the CPUID in CPU template

### DISCUSSION THREAD

**Discussion by up2wing** [firecracker-microvm/firecracker#4256] [Q&A] [answered]
Hi All,
I'm newbi in firecracker, many thx for the excellent work!

Although I know the static CPU template will be depracated from v1.5.0, however I have some puzzles want to clarify, and If someone can help, I will appreciate very much.

> Intel T2CL and AMD T2A templates together aim to provide instruction set feature parity between CPUs running them, so they can form a heterogeneous fleet exposing the same instruction sets to the application.
It's easy to understand to filter the CPUID entry if one of CascadeLake/Milan doesn't have, however, after looking through the source code in t2cl.rs, I found some entry is ignored, such like ADX(0x00000007 0x00: EBX [19]), but CascadeLake/Milan both support this feature[1][2][3]. Why does firecracker filter this feature?

These are the similar CPUID which are filtered:
0x00000007 0x00: EBX [18] RDSEED
0x00000007 0x00: EBX [19] ADX
0x00000007 0x00: EBX [24] CLWB 
0x00000007 0x00: EBX [23] CLFLUSHOPT 
0x00000007 0x00: ECX [02] UMIP 
0x00000007 0x00: ECX [03] PKU 
0x0000000d 0x01: EAX [01] XSAVEC 

Any comment or help is highly appreciated!

1. https://en.wikichip.org/wiki/intel/microarchitectures/skylake_(server)
2. https://en.wikichip.org/wiki/amd/cores/milan
3. https://en.wikichip.org/wiki/intel/cores/cascade_lake_ap

**Comment by kalyazin** [comment, marked as answer]
Hi @up2wing, thanks for starting a discussion!

The reason we decided against including RDSEED, ADX, CLWB and CLFLUSHOPT bits in the T2CL/T2A CPUs templates was that they did not have different values between Cascade Lake and Milan.

The reason we did not include UMIP, PKU and XSAVEC was regular userspace applications (that Firecracker is mainly targeted at) were not supposed to observe the effects that settings those bits would make. We limited those CPU templates to feature bits that are likely to be taken into account by applications or frameworks that choose different (optimised) implementations based on them.

Also, as a note, we have [deprecated](https://github.com/firecracker-microvm/firecracker/blob/main/CHANGELOG.md#changed-1) static CPU templates in Firecracker 1.5. Users are encouraged to use [custom CPU templates](https://github.com/firecracker-microvm/firecracker/blob/main/docs/cpu_templates/cpu-templates.md#custom-cpu-templates) instead where they have much more control over the guest CPUID configuration.

Hope it answers your question.

**Reply by up2wing** [reply]
This address my problem. Many thanks!
